tfhe/src/core_crypto/commons/math/random/tests.rs

```diff
@@ -504,17 +547,152 @@ fn test_uniform_random_custom_mod<
+        _ciphertext_modulus: CiphertextModulus<Scalar>,
+    ) -> Scalar {
+        Scalar::cast_from(input)
+    }
+
+    fn map_value_to_usize(
+        &self,
+        input: Scalar,
+        _ciphertext_modulus: CiphertextModulus<Scalar>,
+    ) -> usize {
+        input.cast_into()
+    }
+}
+
 // This test takes care of all bigger native types, as underneath the CSPRNG outputs bytes
 #[test]
 fn test_uniform_random_native_mod_u8() {
-    test_uniform_random_custom_mod::<u8>(CiphertextModulus::new_native());
+    let ciphertext_modulus = CiphertextModulus::new_native();
+    test_random_from_distribution_custom_mod::<u8, Uniform>((), ciphertext_modulus);
 }
 
 #[test]
 fn test_uniform_random_custom_mod_u64() {
     // Have a modulus that will generate some rejection but is relatively small to be able to test
     // quickly
-    test_uniform_random_custom_mod::<u64>(
-        CiphertextModulus::try_new((1 << 10) + (1 << 9)).unwrap(),
-    );
+    let ciphertext_modulus = CiphertextModulus::try_new((1 << 10) + (1 << 9)).unwrap();
+    test_random_from_distribution_custom_mod::<u64, Uniform>((), ciphertext_modulus);
+}
+
+impl<Scalar: UnsignedInteger + CastFrom<usize> + CastInto<usize>> DistributionTestHelper<Scalar>
+    for TUniform<Scalar>
+{
+    type CreationInfos = u32;
+
+    fn new_with_custom_modulus(
+        value: Self::CreationInfos,
+        ciphertext_modulus: CiphertextModulus<Scalar>,
+    ) -> Self {
+        assert!(ciphertext_modulus.is_native_modulus());
+        Self::new(value)
+    }
+
+    fn distinct_values(&self, _ciphertext_modulus: CiphertextModulus<Scalar>) -> usize {
+        // TUniform does not support non native moduli for now
+        self.distinct_value_count()
+    }
+
+    fn cumulative_distribution_function(
+        &self,
+        integer_value: Scalar,
+        _distinct_values: usize,
+        ciphertext_modulus: CiphertextModulus<Scalar>,
+    ) -> f64 {
+        let max_value_inclusive = self.max_value_inclusive();
+        let integer_value_signed: Scalar::Signed = integer_value.cast_into();
+        let value_index: usize = self.map_value_to_usize(integer_value, ciphertext_modulus);
+        // CDF for the TUniform distribution
+        if integer_value_signed == max_value_inclusive {
+            1.0
+        } else {
+            2.0f64.powi(-(self.bound_log2() as i32 + 2))
+                + 2.0f64.powi(-(self.bound_log2() as i32 + 1)) * value_index as f64
+        }
+    }
+
+    fn map_usize_to_value(
+        &self,
+        input: usize,
+        _ciphertext_modulus: CiphertextModulus<Scalar>,
+    ) -> Scalar {
+        // Input is in [0; 2^(b + 1)]
+        let input_as_scalar = Scalar::cast_from(input);
+        // Cast to signed
+        let input_as_signed_scalar: Scalar::Signed = input_as_scalar.cast_into();
+        let min_value_inclusive = self.min_value_inclusive();
+        // This is in [-2^b; 2^b]
+        let value_as_signed = input_as_signed_scalar + min_value_inclusive;
```

Comment:
Could be a wrapping add

Reply:
hmm why though ?

Reply:
In case `b=Scalar::BITS`
Not sure if this is supported

Reply:
it's not, the new function of TUniform forbids it, it only allows b <= Scalar::BITS - 2 to be able to distinguish between -2^b and 2^b because of 2's complement